MapboxCoreNavigation/RouteController.swift

```diff
@@ -672,22 +677,30 @@ extension RouteController: CLLocationManagerDelegate {
     }
     
     func detectRouteStepInTunnel(for location: CLLocation) {
-        guard routeProgress.currentLegProgress.currentStep.containsTunnel, let coordinates = routeProgress.currentLegProgress.currentStep.coordinates else { return }
-        guard let tunnelSlice = routeProgress.currentLegProgress.currentStep.tunnelSlice else { return }
-        guard let tunnelStartCoordinate = tunnelSlice.coordinates.first, let tunnelEndCoordinate = tunnelSlice.coordinates.last else { return }
+        guard let intersectionBounds = routeProgress.currentLegProgress.currentStep.tunnelIntersectionsBounds, let coordinates = routeProgress.currentLegProgress.currentStep.coordinates else { return }
         
-        // Calculated distances from current location to tunnel entrance and exit
-        let tunnelDistance = tunnelSlice.distance()
-        let distanceToEntrance = Polyline(coordinates).distance(from: location.coordinate, to: tunnelStartCoordinate)
-        let distanceToExit = Polyline(coordinates).distance(from: location.coordinate, to: tunnelEndCoordinate)
-
-        if distanceToEntrance <= tunnelDistance && distanceToExit <= tunnelDistance {
-            delegate?.routeController?(self, didEnterTunnel: true)
+        if hasEnteredTunnel(intersectionBounds, coordinates: coordinates) {
+            delegate?.routeControllerDidEnterTunnel?(self)
         } else {
-            delegate?.routeController?(self, didEnterTunnel: false)
+            delegate?.routeControllerDidExitTunnel?(self)
         }
     }
     
+    func hasEnteredTunnel(_ intersectionBounds: [IntersectionBounds], coordinates: [CLLocationCoordinate2D]) -> Bool {
+        guard let startLocation = coordinates.first else { return false }
+        
+        let distanceTraveled = routeProgress.distanceTraveled
+        for bounds in intersectionBounds {
+            let distanceToTunnelEntrance = Polyline(coordinates).distance(from: startLocation, to: bounds.entry.location)
+            let distanceToTunnelExit = Polyline(coordinates).distance(from: startLocation, to: bounds.exit.location)
```

Comment:
On each location update, this turns the step’s coordinates into a polyline (which has a memory cost) and computes a distance along the polyline (not cheap) twice for each tunnel along the route – more if there are highway exits and entrances within the tunnel.

Out of curiosity, did you find any problems with the approach suggested in https://github.com/mapbox/mapbox-navigation-ios/pull/1127#discussion_r169512237? That approach would’ve relied on storing the just-passed intersection on the route step progress, similar to `distanceTraveled`, making the current road segment’s classes much more accessible, and in _O_(1) time.

The suggestion in https://github.com/mapbox/mapbox-navigation-ios/pull/1127#discussion_r169518181 was to iterate over _all_ the step’s intersections, gathering their distances from the beginning of the step, and to do it just once, as soon as the user begins the step.